// butter-settings-popcorntime.app/index.js
module.exports = {
    projectName: 'Popcorn Time',
    projectUrl: 'https://popcorntime.app',
    projectTwitter: 'popcorntimetv',
    projectTwitter2: 'PopcorntimeappR',
    projectBlog: 'https://blog.popcorntime.app/',
    projectForum: 'https://www.reddit.com/r/PopcornTime',
    projectForum2: 'https://discuss.popcorntime.app',
    projectForum3: 'https://www.reddit.com/r/PopcornTimeApp',
    statusUrl: 'http://status.popcorntime.app',
    changelogUrl: 'https://github.com/popcorn-official/popcorn-desktop/commits/master',
    issuesUrl: 'https://github.com/popcorn-official/popcorn-desktop/issues',
    sourceUrl: 'https://github.com/popcorn-official/popcorn-desktop/',
    commitUrl: 'https://github.com/popcorn-official/popcorn-desktop/commit',
    providers: {
        movie: {
            order: 1,
            name: 'Movies',
            uri: ['MovieApi?'
                  +'&apiURL='
                    + 'https://movies-v2.api-fetch.sh/,'
                    + 'https://movies-v2.api-fetch.am/,'
                   +  'https://movies-v2.api-fetch.website/'
                 ]
        },
        tvshow: {
            order: 2,
            name: 'Series',
            uri: ['TVApi?'
                  +'&apiURL='
                    + 'https://tv-v2.api-fetch.sh/,'
                    + 'https://tv-v2.api-fetch.am/,'
                    + 'https://tv-v2.api-fetch.website/'
                 ]
        },
        anime: {
            order: 3,
            name: 'Anime',
            uri: ['AnimeApi?'
                  +'&apiURL='
                    + 'https://anime.api-fetch.sh/,'
                    + 'https://anime.api-fetch.am/,'
                    + 'https://anime.api-fetch.website/'
                 ]
        },
        subtitle: 'OpenSubtitles',
        metadata: 'Trakttv',
        tvst: 'TVShowTime',
        torrentCache: 'TorrentCache'
    },
    customServers: {
        movie : [
            'https://yts.mx/',
            'https://popcorn-ru.tk/'
        ],
        tvshow: [
            'https://popcorn-ru.tk/'
        ],
        anime: [
        ]
    },
    updateEndpoint: {
        url: 'https://popcorntime.app/',
        index: 0,
        proxies: [{
            url: 'https://popcorntime.app/'
        }]
    },
    updateKey: '-----BEGIN PUBLIC KEY-----\n' +
        'MIIBtjCCASsGByqGSM44BAEwggEeAoGBAPNM5SX+yR8MJNrX9uCQIiy0t3IsyNHs\n' +
        'HWA180wDDd3S+DzQgIzDXBqlYVmcovclX+1wafshVDw3xFTJGuKuva7JS3yKnjds\n' +
        'NXbvM9CrJ2Jngfd0yQPmSh41qmJXHHSwZfPZBxQnspKjbcC5qypM5DqX9oDSJm2l\n' +
        'fM/weiUGnIf7AhUAgokTdF7G0USfpkUUOaBOmzx2RRkCgYAyy5WJDESLoU8vHbQc\n' +
        'rAMnPZrImUwjFD6Pa3CxhkZrulsAOUb/gmc7B0K9I6p+UlJoAvVPXOBMVG/MYeBJ\n' +
        '19/BH5UNeI1sGT5/Kg2k2rHVpuqzcvlS/qctIENgCNMo49l3LrkHbJPXKJ6bf+T2\n' +
        '8lFWRP2kVlrx/cHdqSi6aHoGTAOBhAACgYBTNeXBHbWDOxzSJcD6q4UDGTnHaHHP\n' +
        'JgeCrPkH6GBa9azUsZ+3MA98b46yhWO2QuRwmFQwPiME+Brim3tHlSuXbL1e5qKf\n' +
        'GOm3OxA3zKXG4cjy6TyEKajYlT45Q+tgt1L1HuGAJjWFRSA0PP9ctC6nH+2N3HmW\n' +
        'RTcms0CPio56gg==\n' +
        '-----END PUBLIC KEY-----\n',
    httpApiUsername: 'popcorn',
    httpApiPassword: 'popcorn',
    activateVpn: false
}
